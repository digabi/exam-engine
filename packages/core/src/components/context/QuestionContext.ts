import React from 'react'
import { findChildElement, getNumericAttribute, parentElements, queryAll } from '../../dom-utils'
import { withContext } from './withContext'
import { ExamComponentProps } from '../../createRenderChildNodes'

export interface QuestionContext {
  answers: Element[]
  displayNumber: string
  hasExternalMaterial: boolean
  maxAnswers?: number
  maxScore: number
  level: number
  childQuestions: Element[]
}

export const QuestionContext = React.createContext<QuestionContext>({} as QuestionContext)

export const withQuestionContext = withContext<QuestionContext, ExamComponentProps>(QuestionContext, ({ element }) => {
  const childQuestions = queryAll(element, 'question', false)
  const answers = childQuestions.length
    ? []
    : queryAll(element, ['text-answer', 'scored-text-answer', 'dropdown-answer', 'choice-answer'], false)

  return {
    answers,
    displayNumber: element.getAttribute('display-number')!,
    hasExternalMaterial: findChildElement(element, 'external-material') != null,
    maxAnswers: getNumericAttribute(element, 'max-answers'),
    maxScore: getNumericAttribute(element, 'max-score')!,
    level: parentElements(element, 'question').length,
    childQuestions,
  }
})
