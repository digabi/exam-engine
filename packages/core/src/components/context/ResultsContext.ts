import * as _ from 'lodash-es'
import React from 'react'
import { findChildrenAnswers, getNumericAttribute, parentElements, queryAll } from '../../dom-utils'
import { ChoiceAnswer, ChoiceGroupChoice, ExamAnswer, GradingStructure, QuestionId, Score } from '../../index'
import { withContext } from './withContext'
import { ResultsProps } from '../results/Results'

export interface ResultsContext {
  answersByQuestionId: Record<QuestionId, ExamAnswer>
  gradingStructure: GradingStructure
  scores: Score[]
  gradingText: string | undefined
  totalScore: number
  singleGrading?: boolean
}

export const ResultsContext = React.createContext<ResultsContext>({} as ResultsContext)

export const withResultsContext = withContext<ResultsContext, ResultsProps>(
  ResultsContext,
  ({ scores, doc, gradingStructure, gradingText, answers, singleGrading }) => {
    const answersByQuestionId = _.keyBy(answers, 'questionId')
    const topLevelQuestions = queryAll(doc.documentElement, 'question', false)
    const totalScore = calculateQuestionsTotalSumScore(topLevelQuestions, gradingStructure, scores, answersByQuestionId)

    return {
      answersByQuestionId,
      gradingStructure,
      scores,
      totalScore,
      gradingText,
      singleGrading
    }
  }
)

export function findMultiChoiceFromGradingStructure(
  gradingStructure: GradingStructure,
  id: number
): ChoiceGroupChoice | undefined {
  const choiceGroups = gradingStructure?.questions.filter(v => v.type === 'choicegroup') || []

  for (let i = 0, length = choiceGroups.length; i < length; i++) {
    const choiceGroup = choiceGroups[i]
    for (let j = 0, choicesLength = choiceGroup.choices.length; j < choicesLength; j++) {
      if (choiceGroup.choices[j].id === id) {
        return choiceGroup.choices[j]
      }
    }
  }
  return undefined
}

export function findScore(scores: Score[], questionId: number): Score | undefined {
  return scores.find(a => a.questionId === questionId)
}

export function calculateQuestionsTotalSumScore(
  topLevelQuestions: Element[],
  gradingStructure: GradingStructure,
  scores: Score[],
  answersByQuestionId: Record<QuestionId, ExamAnswer>
): number {
  return _.sum(
    topLevelQuestions.map(question =>
      calculateQuestionSumScore(question, gradingStructure, scores, answersByQuestionId)
    )
  )
}

export function calculateQuestionSumScore(
  questionElement: Element,
  gradingStructure: GradingStructure,
  scores: Score[],
  answersById: Record<QuestionId, ExamAnswer>
): number {
  const choiceQuestionScore = (questionId: number, scoredAnswer: ChoiceAnswer) => {
    const choice = findMultiChoiceFromGradingStructure(gradingStructure, questionId)
    return choice ? choice.options.find(o => o.id === Number(scoredAnswer.value))?.score : 0
  }

  const textQuestionScore = (questionId: number) => {
    const score = findScore(scores, questionId)
    return score
      ? [score.inspection, score.censoring?.scores[0], score.autograding, score.pregrading].find(Boolean)?.score || 0
      : 0
  }

  const sumScore = _.sum(
    findChildrenAnswers(questionElement).map(answer => {
      const questionId = getNumericAttribute(answer, 'question-id')!
      const scoredAnswer = answersById[questionId]
      if (!scoredAnswer) {
        return 0
      }
      if (scoredAnswer.type === 'choice') {
        return choiceQuestionScore(questionId, scoredAnswer)
      } else {
        return textQuestionScore(questionId)
      }
    })
  )
  // top level questions may not yield negative scores
  if (parentElements(questionElement, 'question').length === 0 && sumScore < 0) {
    return 0
  }
  return sumScore
}
