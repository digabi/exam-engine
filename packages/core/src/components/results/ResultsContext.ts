import { ChoiceGroupChoice, ChoiceGroupQuestion, GradingStructure } from '@digabi/exam-engine-mastering'
import * as _ from 'lodash-es'
import React from 'react'
import { findChildrenAnswers, getNumericAttribute, parentElements, queryAll } from '../../dom-utils'
import { AnswerScore, ChoiceAnswer, ExamAnswer, QuestionId } from '../types'
import { withContext } from '../withContext'
import { ResultsProps } from './Results'

export interface ResultsContext {
  answersByQuestionId: Record<QuestionId, ExamAnswer>
  gradingStructure: GradingStructure
  scores: AnswerScore[]
  gradingText: string | undefined
  totalScore: number
}

export const ResultsContext = React.createContext<ResultsContext>({} as ResultsContext)

export const withResultsContext = withContext<ResultsContext, ResultsProps>(
  ResultsContext,
  ({ scores, doc, gradingStructure, gradingText, answers }) => {
    const answersByQuestionId = _.keyBy(answers, 'questionId')
    const topLevelQuestions = queryAll(doc.documentElement, 'question', false)
    const totalScore = _.sum(
      topLevelQuestions.map(question =>
        calculateQuestionSumScore(question, gradingStructure, scores, answersByQuestionId)
      )
    )

    return {
      answersByQuestionId,
      gradingStructure,
      scores,
      totalScore,
      gradingText
    }
  }
)

export function findMultiChoiceFromGradingStructure(
  gradingStructure: GradingStructure,
  id: number
): ChoiceGroupChoice | undefined {
  const choiceGroups = gradingStructure.questions.filter(v => v.type === 'choicegroup')

  for (let i = 0, length = choiceGroups.length; i < length; i++) {
    const choiceGroup = choiceGroups[i] as ChoiceGroupQuestion
    for (let j = 0, choicesLength = choiceGroup.choices.length; j < choicesLength; j++) {
      if (choiceGroup.choices[j].id === id) {
        return choiceGroup.choices[j]
      }
    }
  }
  return undefined
}

export function findScore(scores: AnswerScore[], questionId: number) {
  return scores.find(s => s.questionId === questionId)
}

export function calculateQuestionSumScore(
  questionElement: Element,
  gradingStructure: GradingStructure,
  scores: AnswerScore[],
  answersById: Record<QuestionId, ExamAnswer>
) {
  const choiceQuestionScore = (questionId: number, scoredAnswer: ChoiceAnswer) => {
    const choice = findMultiChoiceFromGradingStructure(gradingStructure, questionId)
    return choice ? choice.options.find(o => o.id === Number(scoredAnswer.value))!.score : 0
  }

  const textQuestionScore = (questionId: number) => {
    const score = findScore(scores, questionId)
    return score ? score.scoreValue : 0
  }

  const sumScore = _.sum(
    findChildrenAnswers(questionElement).map(answer => {
      const questionId = getNumericAttribute(answer, 'question-id')!
      const scoredAnswer = answersById[questionId]
      if (!scoredAnswer) {
        return 0
      }
      if (scoredAnswer.type === 'choice') {
        return choiceQuestionScore(questionId, scoredAnswer)
      } else {
        return textQuestionScore(questionId)
      }
    })
  )
  // top level questions may not yield negative scores
  if (parentElements(questionElement, 'question').length === 0 && sumScore < 0) {
    return 0
  }
  return sumScore
}
