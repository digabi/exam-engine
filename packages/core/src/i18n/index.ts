import i18n from 'i18next'
import * as _ from 'lodash-es'
import { A_E } from './A_E'
import { BA } from './BA'
import { CA } from './CA'
import { EA } from './EA'
import { FA } from './FA'
import { fi_FI } from './fi-FI'
import { GC } from './GC'
import { I } from './I'
import { M } from './M'
import { O_E } from './O_E'
import { PA } from './PA'
import { SA } from './SA'
import { sv_FI } from './sv-FI'
import { TC } from './TC'
import { VA } from './VA'
import { Z } from './Z'

const examSpecificTranslations = {
  A_E,
  BA,
  BB: BA,
  CA,
  CB: CA,
  EA,
  EC: EA,
  FA,
  FC: FA,
  GC,
  I,
  M,
  N: M,
  O_E,
  SA,
  SC: SA,
  PA,
  PC: PA,
  TC,
  VA,
  VC: VA,
  Z,
}

const resources = {
  'fi-FI': {
    translation: fi_FI,
    ...examSpecificTranslations,
  },
  'sv-FI': {
    translation: sv_FI,
    ...examSpecificTranslations,
  },
}

export function initI18n(language: string, examCode: string | null, dayCode: string | null): typeof i18n {
  const namespace = examCode ? examCode + (dayCode ? `_${dayCode}` : '') : 'translation'

  return i18n.createInstance(
    {
      resources,
      lng: language,
      fallbackLng: 'fi-FI',
      debug: false,
      defaultNS: namespace,
      fallbackNS: 'translation',
      interpolation: {
        escapeValue: false,
        format: (value, format) => {
          switch (format) {
            case 'range': {
              // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
              const [start, end] = value
              // eslint-disable-next-line @typescript-eslint/no-unsafe-return,@typescript-eslint/restrict-template-expressions
              return start == null ? end : end == null ? start : start === end ? start : `${start}â€“${end}`
            }
            case 'first':
              return _.first(value)
            case 'last':
              return _.last(value)
            default:
              // eslint-disable-next-line @typescript-eslint/no-unsafe-return
              return value
          }
        },
      },
    },
    _.noop
  )
}

export const changeLanguage = (_i18n: typeof i18n, language: string) => (): void => {
  if (_i18n.language !== language) {
    _i18n.changeLanguage(language).then(_.noop).catch(_.noop)
  }
}

export default i18n
