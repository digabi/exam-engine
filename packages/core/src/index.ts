import '../src/css/main.less'
import Attachments from './components/attachments/Attachments'
import Exam from './components/exam/Exam'
import GradingInstructions from './components/grading-instructions/GradingInstructions'
import Results from './components/results/Results'
import parseExam from './parser/parseExam'

export { Attachments, Exam, GradingInstructions, Results, parseExam }
export { ExamComponentProps } from './createRenderChildNodes'
export * from './types/ExamAnswer'
export * from './types/ExamServerAPI'
export * from './types/GradingStructure'
export * from './types/Score'

export interface ExamBundle {
  Attachments: typeof Attachments
  Exam: typeof Exam
  parseExam: typeof parseExam
  Results: typeof Results
}
