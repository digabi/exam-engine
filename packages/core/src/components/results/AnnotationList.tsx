import * as _ from 'lodash-es'
import React, { useContext } from 'react'
import { useTranslation } from 'react-i18next'
import { getNumericAttribute } from '../../dom-utils'
import { shortDisplayNumber } from '../../shortDisplayNumber'
import { mapMaybe } from '../../utils'
import { QuestionContext } from '../QuestionContext'
import { findScore, ResultsContext } from './ResultsContext'
import { Annotation, Score } from '../..'

interface AnnotationItem {
  numbering: string
  message: string
}

interface AnnotationListProps {
  i18nTitleKey?: string
  annotations?: AnnotationItem[]
}

const hasAnnotations = (score: Score) =>
  Boolean(score?.pregrading?.annotations?.length || score?.censoring?.annotations?.length)

const getPrefix = (answers: Element[], answer: Element) =>
  answers.length > 1 ? shortDisplayNumber(answer.getAttribute('display-number')!) : ''

const AnnotationListComponent = ({ i18nTitleKey, annotations }: AnnotationListProps) => {
  const { t } = useTranslation()

  return annotations ? (
    <>
      {i18nTitleKey && <h5>{t(i18nTitleKey)}</h5>}
      <ol className="e-list-data e-pad-l-0 e-font-size-s">
        {annotations.map(({ numbering, message }) => (
          <li data-list-number={numbering} key={numbering}>
            {message}
          </li>
        ))}
      </ol>
    </>
  ) : null
}

function ResultsAnnotationList() {
  const { answers } = useContext(QuestionContext)
  const { scores, singleGrading } = useContext(ResultsContext)

  const answersAndScores = mapMaybe(answers, (answer) => {
    const questionId = getNumericAttribute(answer, 'question-id')
    const score = findScore(scores, questionId!)

    return score && answer && hasAnnotations(score) ? ([answer, score] as const) : undefined
  })

  const getListOfAnnotations = (
    answerElementAndScores: Array<readonly [Element, Score]>,
    annotationsFrom: 'pregrading' | 'censoring',
    listNumberOffset = 0
  ) =>
    _.flatMap(
      answerElementAndScores,
      ([answer, score]) =>
        score[annotationsFrom]?.annotations
          ?.filter((a) => a.message.length)
          .map((annotation: Annotation, i: number) => {
            const numbering = getPrefix(answers, answer) + String(listNumberOffset + i + 1) + ')'
            const message = annotation.message
            return { numbering, message }
          }) ?? []
    )

  const pregradingAnnotations = getListOfAnnotations(answersAndScores, 'pregrading')
  const censoringAnnotations = getListOfAnnotations(answersAndScores, 'censoring', pregradingAnnotations.length)

  return pregradingAnnotations.length || censoringAnnotations.length ? (
    <div className="e-annotation-list e-columns e-mrg-t-2">
      {singleGrading ? (
        <div className="e-column e-column--10">
          <AnnotationListComponent annotations={pregradingAnnotations} />
        </div>
      ) : (
        <>
          <div className="e-column e-column--6">
            <AnnotationListComponent
              i18nTitleKey={'grading.pregrading-annotations'}
              annotations={pregradingAnnotations}
            />
          </div>
          <div className="e-column e-column--6">
            <AnnotationListComponent i18nTitleKey={'grading.censor-annotations'} annotations={censoringAnnotations} />
          </div>
        </>
      )}
    </div>
  ) : null
}

export default React.memo(ResultsAnnotationList)
