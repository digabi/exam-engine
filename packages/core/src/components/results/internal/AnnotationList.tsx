import * as _ from 'lodash-es'
import React, { useContext } from 'react'
import { Annotation, Score } from '../../../index'
import { getNumericAttribute } from '../../../dom-utils'
import { shortDisplayNumber } from '../../../shortDisplayNumber'
import { mapMaybe } from '../../../utils'
import { QuestionContext } from '../../context/QuestionContext'
import { findScore, ResultsContext } from '../../context/ResultsContext'
import { AnnotationLists } from '../../shared/AnnotationLists'

const hasAnnotations = (score: Score) =>
  Boolean(score?.pregrading?.annotations?.length || score?.censoring?.annotations?.length)

const getPrefix = (answers: Element[], answer: Element) =>
  answers.length > 1 ? shortDisplayNumber(answer.getAttribute('display-number')!) : ''

function ResultsAnnotationList() {
  const { answers } = useContext(QuestionContext)
  const { scores, singleGrading } = useContext(ResultsContext)

  const answersAndScores = mapMaybe(answers || [], answer => {
    const questionId = getNumericAttribute(answer, 'question-id')
    const score = findScore(scores, questionId!)

    return score && answer && hasAnnotations(score) ? ([answer, score] as const) : undefined
  })

  const getListOfAnnotations = (
    answerElementAndScores: Array<readonly [Element, Score]>,
    annotationsFrom: 'pregrading' | 'censoring',
    listNumberOffset = 0
  ) =>
    _.flatMap(
      answerElementAndScores,
      ([answer, score]) =>
        score[annotationsFrom]?.annotations
          ?.filter(a => ('length' in a ? !!a.length : true))
          .map((annotation: Annotation, i: number) => {
            const numbering = `${getPrefix(answers, answer) + String(listNumberOffset + i + 1)})`
            const message = annotation.message
            return { numbering, message }
          }) ?? []
    )

  const pregradingAnnotations = getListOfAnnotations(answersAndScores, 'pregrading')
  const censoringAnnotations = getListOfAnnotations(answersAndScores, 'censoring', pregradingAnnotations.length)

  return (
    <AnnotationLists
      pregradingAnnotations={pregradingAnnotations}
      censoringAnnotations={censoringAnnotations}
      singleGrading={!!singleGrading}
    />
  )
}

export default React.memo(ResultsAnnotationList)
