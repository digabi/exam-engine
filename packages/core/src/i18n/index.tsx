import i18n, { PostProcessorModule, TOptions } from 'i18next'
import * as _ from 'lodash-es'
import React from 'react'
// eslint-disable-next-line no-restricted-imports
import { TFunction, useTranslation } from 'react-i18next'
import { A_E } from './A_E'
import { BA } from './BA'
import { CA } from './CA'
import { EA } from './EA'
import { FA } from './FA'
import { fi_FI } from './fi-FI'
import { GC } from './GC'
import { I } from './I'
import { M } from './M'
import { O_E } from './O_E'
import { PA } from './PA'
import { SA } from './SA'
import { sv_FI } from './sv-FI'
import { TC } from './TC'
import { VA } from './VA'
import { Z } from './Z'

const examSpecificTranslations = {
  A_E,
  BA,
  BB: BA,
  CA,
  CB: CA,
  EA,
  EC: EA,
  FA,
  FC: FA,
  GC,
  I,
  M,
  N: M,
  O_E,
  SA,
  SC: SA,
  PA,
  PC: PA,
  TC,
  VA,
  VC: VA,
  Z,
}

const resources = {
  'fi-FI': {
    translation: fi_FI,
    ...examSpecificTranslations,
  },
  'sv-FI': {
    translation: sv_FI,
    ...examSpecificTranslations,
  },
}

declare module 'react-i18next' {
  type ExamEngineI18nResources = { translation: typeof fi_FI }

  // eslint-disable-next-line @typescript-eslint/no-empty-interface
  interface Resources extends ExamEngineI18nResources {}
}

export function initI18n(language: string, examCode?: string, dayCode?: string): typeof i18n {
  const namespace = examCode ? examCode + (dayCode ? `_${dayCode}` : '') : 'translation'

  return i18n
    .use<PostProcessorModule>({
      type: 'postProcessor',
      name: 'lowercase',
      process: (value, _key, options) => value.toLocaleLowerCase(options.lng),
    })
    .createInstance(
      {
        compatibilityJSON: 'v3',
        resources,
        lng: language,
        fallbackLng: 'fi-FI',
        debug: false,
        defaultNS: namespace,
        fallbackNS: 'translation',
        interpolation: {
          escapeValue: false,
          format: (value, format) => {
            const val = value as [string, string]
            switch (format) {
              case 'range': {
                const [start, end] = val
                return start == null ? end : end == null ? start : start === end ? start : `${start}–${end}`
              }
              case 'first':
                return _.first(val)
              case 'last':
                return _.last(val)
              default:
                // eslint-disable-next-line @typescript-eslint/no-unsafe-return
                return value
            }
          },
        },
      },
      _.noop
    )
}

export function useExamTranslation() {
  const { i18n, t: originalT } = useTranslation()
  type Translator = {
    (key: string | string[], options?: TOptions): string | JSX.Element
    raw: TFunction
  }
  const t: Translator = (key, options) => translate(i18n, key, options)
  t.raw = originalT
  return { i18n, t }
}

/**
 * Custom wrapper for i18n.t that wraps the translated text in a span element
 * containing the correct lang attribute, if necessary.
 *
 * If you want to disable the wrapping, you can use `t.raw(…)`.
 */
function translate(instance: typeof i18n, key: string | string[], options?: TOptions): string | JSX.Element {
  const translation = instance.t(key, options)

  const namespace = instance.options.defaultNS!.toString()
  const examHasCustomTranslations = namespace !== 'translation' && namespace in examSpecificTranslations
  const hasCustomTranslationForKey = examHasCustomTranslations && hasTranslation(instance, namespace, key)
  return hasCustomTranslationForKey || !examHasCustomTranslations ? (
    translation
  ) : (
    <span lang={instance.language}>{translation}</span>
  )
}

/**
 * A hacky way to check if we have a translation in the specified namespace.
 * Ideally, we'd probably want to use `i18n.exists()` for this purpose, but I
 * couldn't find a way to disable the fallback mechanism when using it.
 */
function hasTranslation(instance: typeof i18n, namespace: string, key: string | string[]): boolean {
  return Array.isArray(key)
    ? key.some((k) => instance.getResource(instance.language, namespace, k) != null)
    : instance.getResource(instance.language, namespace, key) != null
}

export const changeLanguage = (_i18n: typeof i18n, language: string) => (): void => {
  if (_i18n.language !== language) {
    _i18n.changeLanguage(language).then(_.noop).catch(_.noop)
  }
}

export default i18n
