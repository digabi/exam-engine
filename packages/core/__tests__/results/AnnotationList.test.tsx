import * as i18next from 'i18next'
import _ from 'lodash'
import React from 'react'
import { I18nextProvider } from 'react-i18next'
import { create } from 'react-test-renderer'
import { Score } from '../../src'
import { QuestionContext } from '../../src/components/context/QuestionContext'
import AnnotationList from '../../src/components/results/internal/AnnotationList'
import { ResultsContext } from '../../src/components/context/ResultsContext'
import { initI18n } from '../../src/i18n'

function mkTextAnswer(questionId: number, displayNumber: string, maxScore: number): Element {
  const textAnswer = document.createElement('text-answer')
  textAnswer.setAttribute('type', 'rich-text')
  textAnswer.setAttribute('max-score', String(maxScore))
  textAnswer.setAttribute('question-number', displayNumber)
  textAnswer.setAttribute('question-id', String(questionId))

  return textAnswer
}

const defaultScores: Score = {
  questionId: 1,
  answerId: 1,
  pregrading: {
    score: 1,
    annotations: [
      {
        startIndex: 0,
        length: 1,
        message: 'Test pregrading annotation',
      },
    ],
  },
  censoring: {
    scores: [
      { score: 4, shortCode: 'SE3' },
      { score: 3, shortCode: 'SE2' },
      { score: 2, shortCode: 'SE1' },
    ],
    annotations: [
      {
        startIndex: 0,
        length: 1,
        message: 'Test censoring annotation',
      },
    ],
  },
  inspection: { score: 5, shortCodes: ['IN1', 'IN2'] },
}

const defaultAnswer = mkTextAnswer(1, '5.1', 12)

describe('<AnnotationList />', () => {
  let i18n: i18next.i18n

  describe('fi-FI', () => {
    beforeAll(() => {
      i18n = initI18n('fi-FI')
    })

    it('renders empty without score', () => {
      const resultsProps = {
        scores: [],
      }
      expect(renderWithContext(resultsProps, [defaultAnswer]).toJSON()).toMatchSnapshot()
    })

    it('renders empty without annotations', () => {
      const resultsProps = {
        scores: [_.pick(defaultScores, 'answerId', 'questionId')],
      }
      expect(renderWithContext(resultsProps, [defaultAnswer]).toJSON()).toMatchSnapshot()
    })

    it('renders with only pregrading annotations', () => {
      const resultsProps = {
        scores: [_.pick(defaultScores, 'pregrading', 'answerId', 'questionId')],
      }
      expect(renderWithContext(resultsProps, [defaultAnswer]).toJSON()).toMatchSnapshot()
    })

    it('filters empty annotations', () => {
      const resultsProps = {
        scores: [
          {
            ..._.pick(defaultScores, 'pregrading', 'answerId', 'questionId'),
            ...{
              pregrading: {
                score: 1,
                annotations: [
                  {
                    startIndex: 0,
                    length: 1,
                    message: 'Before empty annotation',
                  },
                  {
                    startIndex: 1,
                    length: 1,
                    message: '',
                  },
                  {
                    startIndex: 2,
                    length: 1,
                    message: 'After empty annotation',
                  },
                ],
              },
            },
          },
        ],
      }
      expect(renderWithContext(resultsProps, [defaultAnswer]).toJSON()).toMatchSnapshot()
    })

    it('renders with only censoring annotations', () => {
      const resultsProps = {
        scores: [_.pick(defaultScores, 'censoring', 'answerId', 'questionId')],
      }
      expect(renderWithContext(resultsProps, [defaultAnswer]).toJSON()).toMatchSnapshot()
    })

    it('renders with pregrading and censoring annotations', () => {
      const resultsProps = {
        scores: [defaultScores],
      }
      expect(renderWithContext(resultsProps, [defaultAnswer]).toJSON()).toMatchSnapshot()
    })

    it('renders only pregrading without header on singleGrading in ResultContext', () => {
      const resultsProps = {
        scores: [defaultScores],
        singleGrading: true,
      }
      expect(renderWithContext(resultsProps, [defaultAnswer]).toJSON()).toMatchSnapshot()
    })

    it('renders null if no scores and singleGrading in ResultContext', () => {
      const resultsProps = {
        scores: [],
        singleGrading: true,
      }
      expect(renderWithContext(resultsProps, [defaultAnswer]).toJSON()).toMatchSnapshot()
    })
  })

  describe('sv-FI', () => {
    it('renders with pregrading and censoring annotations', () => {
      i18n = initI18n('sv-FI')
      const resultsProps = {
        scores: [defaultScores],
      }
      expect(renderWithContext(resultsProps, [defaultAnswer]).toJSON()).toMatchSnapshot()
    })
  })

  const questionContextProps: QuestionContext = {
    answers: [],
    displayNumber: '1',
    hasExternalMaterial: false,
    maxAnswers: 2,
    maxScore: 2,
    level: 2,
    childQuestions: [],
    questionLabelIds: '',
  }

  const resultsContextDefaults: ResultsContext = {
    answersByQuestionId: {},
    gradingStructure: { questions: [] },
    scores: [],
    gradingText: '',
    totalScore: 60,
  }

  function renderWithContext(results: Partial<ResultsContext>, answers: Element[]) {
    return create(
      <QuestionContext.Provider value={{ ...questionContextProps, ...{ answers } }}>
        <ResultsContext.Provider value={{ ...resultsContextDefaults, ...results }}>
          <I18nextProvider i18n={i18n}>
            <AnnotationList />
          </I18nextProvider>
        </ResultsContext.Provider>
      </QuestionContext.Provider>
    )
  }
})
