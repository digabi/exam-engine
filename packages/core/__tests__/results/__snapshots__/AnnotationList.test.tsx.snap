// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`<AnnotationList /> fi-FI filters empty annotations 1`] = `
<div
  className="e-annotation-list e-columns e-mrg-t-2"
>
  <div
    className="e-column e-column--6"
  >
    <h5>
      Alustavan arvostelun merkinnät (opettaja)
    </h5>
    <ol
      className="e-list-data e-pad-l-0 e-font-size-s"
    >
      <li
        data-list-number="1)"
      >
        Before empty annotation
      </li>
      <li
        data-list-number="2)"
      >
        After empty annotation
      </li>
    </ol>
  </div>
  <div
    className="e-column e-column--6"
  >
    <h5>
      Sensorin merkinnät
    </h5>
    <ol
      className="e-list-data e-pad-l-0 e-font-size-s"
    />
  </div>
</div>
`;

exports[`<AnnotationList /> fi-FI renders empty without annotations 1`] = `null`;

exports[`<AnnotationList /> fi-FI renders empty without score 1`] = `null`;

exports[`<AnnotationList /> fi-FI renders null if no scores and singleGrading in ResultContext 1`] = `null`;

exports[`<AnnotationList /> fi-FI renders only pregrading without header on singleGrading in ResultContext 1`] = `
<div
  className="e-annotation-list e-columns e-mrg-t-2"
>
  <div
    className="e-column e-column--10"
  >
    <ol
      className="e-list-data e-pad-l-0 e-font-size-s"
    >
      <li
        data-list-number="1)"
      >
        Test pregrading annotation
      </li>
    </ol>
  </div>
</div>
`;

exports[`<AnnotationList /> fi-FI renders with only censoring annotations 1`] = `
<div
  className="e-annotation-list e-columns e-mrg-t-2"
>
  <div
    className="e-column e-column--6"
  >
    <h5>
      Alustavan arvostelun merkinnät (opettaja)
    </h5>
    <ol
      className="e-list-data e-pad-l-0 e-font-size-s"
    />
  </div>
  <div
    className="e-column e-column--6"
  >
    <h5>
      Sensorin merkinnät
    </h5>
    <ol
      className="e-list-data e-pad-l-0 e-font-size-s"
    >
      <li
        data-list-number="1)"
      >
        Test censoring annotation
      </li>
    </ol>
  </div>
</div>
`;

exports[`<AnnotationList /> fi-FI renders with only pregrading annotations 1`] = `
<div
  className="e-annotation-list e-columns e-mrg-t-2"
>
  <div
    className="e-column e-column--6"
  >
    <h5>
      Alustavan arvostelun merkinnät (opettaja)
    </h5>
    <ol
      className="e-list-data e-pad-l-0 e-font-size-s"
    >
      <li
        data-list-number="1)"
      >
        Test pregrading annotation
      </li>
    </ol>
  </div>
  <div
    className="e-column e-column--6"
  >
    <h5>
      Sensorin merkinnät
    </h5>
    <ol
      className="e-list-data e-pad-l-0 e-font-size-s"
    />
  </div>
</div>
`;

exports[`<AnnotationList /> fi-FI renders with pregrading and censoring annotations 1`] = `
<div
  className="e-annotation-list e-columns e-mrg-t-2"
>
  <div
    className="e-column e-column--6"
  >
    <h5>
      Alustavan arvostelun merkinnät (opettaja)
    </h5>
    <ol
      className="e-list-data e-pad-l-0 e-font-size-s"
    >
      <li
        data-list-number="1)"
      >
        Test pregrading annotation
      </li>
    </ol>
  </div>
  <div
    className="e-column e-column--6"
  >
    <h5>
      Sensorin merkinnät
    </h5>
    <ol
      className="e-list-data e-pad-l-0 e-font-size-s"
    >
      <li
        data-list-number="2)"
      >
        Test censoring annotation
      </li>
    </ol>
  </div>
</div>
`;

exports[`<AnnotationList /> sv-FI renders with pregrading and censoring annotations 1`] = `
<div
  className="e-annotation-list e-columns e-mrg-t-2"
>
  <div
    className="e-column e-column--6"
  >
    <h5>
      Preliminära bedömningens anteckningar (lärare)
    </h5>
    <ol
      className="e-list-data e-pad-l-0 e-font-size-s"
    >
      <li
        data-list-number="1)"
      >
        Test pregrading annotation
      </li>
    </ol>
  </div>
  <div
    className="e-column e-column--6"
  >
    <h5>
      Censorns anteckningar
    </h5>
    <ol
      className="e-list-data e-pad-l-0 e-font-size-s"
    >
      <li
        data-list-number="2)"
      >
        Test censoring annotation
      </li>
    </ol>
  </div>
</div>
`;
